src/palindrom.js

```diff
@@ -193,319 +187,317 @@ var Palindrom = (function() {
-      get: function() {
-        return useWebSocket;
-      },
-      set: function(newValue) {
-        useWebSocket = newValue;
-
-        if (newValue == false) {
-          if (that._ws) {
-            that._ws.onclose = function() {
-              //overwrites the previous onclose
-              that._ws = null;
-            };
-            that._ws.close();
-          }
-          // define wsUrl if needed
-        } else if (!that.wsUrl) {
-          that.wsUrl = toWebSocketURL(that.remoteUrl.href);
+  class PalindromNetworkChannel {
+    constructor(
+      palindrom,
+      remoteUrl,
+      useWebSocket,
+      onReceive,
+      onSend,
+      onConnectionError,
+      onSocketOpened,
+      onFatalError,
+      onStateChange
+    ) {
+      // TODO(tomalec): to be removed once we will achieve better separation of concerns
+      this.palindrom = palindrom;
+
+      if (typeof window !== 'undefined' && window.location) {
+        this.remoteUrl = new URL(remoteUrl, window.location.href);
+      } else {
+        // in Node, URL is absolute
+        this.remoteUrl = new URL(remoteUrl);
+      }
+
+      onReceive && (this.onReceive = onReceive);
+      onSend && (this.onSend = onSend);
+      onConnectionError && (this.onConnectionError = onConnectionError);
+      onFatalError && (this.onFatalError = onFatalError);
+      onStateChange && (this.onStateChange = onStateChange);
+      onSocketOpened && (this.onSocketOpened = onSocketOpened);
+      this._useWebSocket = useWebSocket || false;
+    }
+    get useWebSocket() {
+      return this._useWebSocket;
+    }
+    set useWebSocket(newValue) {
+      this._useWebSocket = newValue;
+
+      if (newValue == false) {
+        if (this._ws) {
+          this._ws.onclose = () => {
+            //overwrites the previous onclose
+            this._ws = null;
+          };
+          this._ws.close();
         }
-        return useWebSocket;
+        // define wsUrl if needed
+      } else if (!this.wsUrl) {
+        this.wsUrl = toWebSocketURL(this.remoteUrl.href);
       }
-    });
-  }
+      return this.useWebSocket;
+    }
 
-  PalindromNetworkChannel.prototype.establish = function(bootstrap) {
-    establish(this, this.remoteUrl.href, null, bootstrap);
-  };
-  PalindromNetworkChannel.prototype.reestablish = function(pending, bootstrap) {
-    establish(
-      this,
-      this.remoteUrl.href + '/reconnect',
-      JSON.stringify(pending),
-      bootstrap
-    );
-  };
+    establish(bootstrap) {
+      establish(this, this.remoteUrl.href, null, bootstrap);
+    }
 
-  // TODO: auto-configure here #38 (tomalec)
-  function establish(network, url, body, bootstrap) {
-    return network.xhr(url, 'application/json', body, function(res) {
-      bootstrap(res.data);
-      if (network.useWebSocket) {
-        network.webSocketUpgrade(network.onSocketOpened);
+    reestablish(pending, bootstrap) {
+      establish(
+        this,
+        `${this.remoteUrl.href}/reconnect`,
```

Comment:
I'd say it's less intuitive, and 2 chars longer 😎  than good old `this.remoteUrl.href+'/reconnect'`. Unless there is some magic behind it that makes it better than before?

Reply:
For me, template literals are more readable. They [can be faster](https://stackoverflow.com/questions/29055518/are-es6-template-literals-faster-than-string-concatenation). There are [more practical reasons](https://ponyfoo.com/articles/template-literals-strictly-better-strings) to use them without hesitation instead of string concatenation.